### PR TITLE
#1470: Improve test coverage for org.cactoos.collection

### PR DESCRIPTION
**Pull request by andreoss**
Per #1470 
- Improve test based on problems found in #1470

### CODE REVIEW

File: src/test/java/org/cactoos/collection/ImmutableTest.java [hunk 1]
```diff
@@ -48,8 +47,8 @@ void size() {
             "size() must be equals to original",
             new Immutable<>(
                 new ListOf<>(1, 2)
-            ),
-            new HasSize(2)
+            ).size(),
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/ImmutableTest.java:50]: @andreoss Please, use `LengthOf` instead of calling `#size()`.

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 2]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
             )
         ).affirm();
     }
+
+    @Test
+    void behavesAsCollection() {
+        new Assertion<>(
+            "Must behave as collection",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:137]: @andreoss I think we must improve this message to something like `Must behave as a no null collection"` to *emphasize* the no nulls capability. WDYT?

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 3]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
...
+                new ListOf<>(1)
+            ),
+            new BehavesAsCollection<>(1)
+        ).affirm();
+    }
+
+    @Test
+    void hasToString() {
+        new Assertion<>(
+            "Must have correct toString",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:148]: @andress the same above.

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 4]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
...
+    }
+
+    @Test
+    void emptyWhenCleared() {
+        final Collection<Integer> col = new NoNulls<>(
+            new ListOf<>(1, 2, 3)
+        );
+        col.clear();
+        new Assertion<>(
+            "Must be empty after #clear()",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:165]: @andress the same above.

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 5]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
...
+    }
+
+    @Test
+    void whenItemRemoved() {
+        final Collection<Integer> col = new NoNulls<>(
+            new ListOf<>(1, 2, 3)
+        );
+        col.remove(1);
+        new Assertion<>(
+            "Must have item removed",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:178]: @andress the same above.

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 6]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
...
+    }
+
+    @Test
+    void whenItemAdded() {
+        final Collection<Integer> col = new NoNulls<>(
+            new ListOf<>(1, 2)
+        );
+        col.add(3);
+        new Assertion<>(
+            "Must have item added",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:191]: @andress the same above.

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 7]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
...
+            col,
+            new IsEqual<>(new ListOf<>(1, 2, 3))
+        ).affirm();
+    }
+
+    @Test
+    void throwsAtAttemptToAddNull() {
+        new Assertion<>(
+            "Must throw exception when add NULL",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:200]: @andress the same above.

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 8]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
...
+    }
+
+    @Test
+    void removesAll() {
+        final Collection<Integer> col = new NoNulls<>(
+            new ListOf<>(1, 2, 3)
+        );
+        col.removeAll(new ListOf<>(1, null));
+        new Assertion<>(
+            "Must have an item removed",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:218]: @andress the same above.

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 9]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
...
+        ).affirm();
+    }
+
+    @Test
+    void retainsAll() {
+        final Collection<Integer> col = new NoNulls<>(
+            new ListOf<>(1, 2, 3)
+        );
+        new Assertion<>(
+            "Must return false for the same collection",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:230]: @andress the same above.

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 10]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
...
+        ).affirm();
+    }
+
+    @Test
+    void retainsAllWithNulls() {
+        final Collection<Integer> col = new NoNulls<>(
+            new ListOf<>(1, 2, 3)
+        );
+        new Assertion<>(
+            "Must return false for a different collection",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:242]: @andress the same above.

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 11]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
...
+            col.retainAll(new ListOf<>(1, null, 3)),
+            new IsTrue()
+        ).affirm();
+    }
+
+    @Test
+    void throwsAtAttemptToRemoveNull() {
+        new Assertion<>(
+            "Must throw exception when removing NULL",
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:251]: @andress the same above.

File: src/test/java/org/cactoos/collection/NoNullsTest.java [hunk 12]
```diff
@@ -120,4 +130,145 @@ void throwsErrorIfNullInAddAll() {
             )
         ).affirm();
     }
+
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/collection/NoNullsTest.java:133]: @andreoss Please, add a test to check a `new NoNulls(new ListOf<>())` (a list already created empty).

**Review by fabriciofx — CHANGES_REQUESTED** [review]
@andreoss Please, can you take a look?

File: src/test/java/org/cactoos/collection/ImmutableTest.java (hunk 1 above)
Comment (andreoss) [reply src/test/java/org/cactoos/collection/ImmutableTest.java:50]: @fabriciofx `#size()` is used explicitly for coverage, `LengthOf`  has the same problem as `HasSize` (i.e counting elements via `#iterator()`)

File: src/test/java/org/cactoos/collection/NoNullsTest.java (hunk 12 above)
Comment (andreoss) [reply src/test/java/org/cactoos/collection/NoNullsTest.java:133]: @fabriciofx Added

**Review by andreoss — COMMENTED** [review]
@fabriciofx Improved on messages. Please, take another look

**Comment by victornoel** [comment]
@fabriciofx is it ok to be merged?

**Comment by victornoel** [comment]
@rultor merge

**Comment by victornoel** [comment]
@andreoss thank you

**Comment by 0crat** [comment]
Code review was too long (48 days), architects (@victornoel) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@0crat quality good
